### DISCUSSION TITLE
request jingle

### DISCUSSION THREAD

**Discussion by jonikane** [AzuraCast/AzuraCast#4384] [Q&A]
Is it possible to customize a request jingle to play before the requested song?

**Comment by bveenker** [comment]
This is a question that I had as well, seems it gets asked every year or two on Azuracast.com

I asked on the LiquidSoap Discord; stefan wrote up the following solution that works great:

1. First, create a playlist with your leader jingles. Set its type to "Advanced"
2. Next, go to Broadcasting->Edit Liquidsoap Configuration
3. Take note of the new playlists location. Mine looks like this:

`playlist_request_leader = playlist(id="playlist_request_leader",mime_type="audio/x-mpegurl",mode="randomize",reload_mode="watch","/var/azuracast/stations/pop/playlists/playlist_request_leader.m3u")`

so my playlist location is `/var/azuracast/stations/pop/playlists/playlist_request_leader.m3u`

4. Go down to the third edit block, the one before the amplification is applied. Put the following script:

```
# Set path to leader's playlist
playlist_request_leader_modified = "/var/azuracast/stations/pop/playlists/playlist_request_leader.m3u"

# Create empty queue
queue_leader = request.queue()

def if_we_have_request_id(m)
    if string.length(m["request_id"]) > 0 then
        # We found a request_id in the metadata. Add the leader to the queue
        playlist_lines = playlist.files("#{playlist_request_leader_modified}")
        random_track = list.nth(playlist_lines, random.int(min=0,max=list.length(playlist_lines)))
        queue_leader.push(request.create(random_track))
    end
end

# Check for request_id in radio's metadata
radio.on_metadata(if_we_have_request_id)

# Briefly interrupt radio with our leader
radio = fallback(track_sensitive=false, [queue_leader, radio])
```

You'll need to change the playlist_request_leader_modified to your playlist's location you noted in step 3

5. Save changes and restart your station

This does exactly what we want, it plays a random jingle in that playlist before every request. One caveat: since we're hacking around with the radio stream, it has a small side effect in that the recently played songs list looks a little weird. It'll show the requested song played twice with the jingle in between. This is due to the fact that the intended song plays briefly, is interrupted by the leader, then starts playing again. Still, small price to pay for this feature. I absolutely love hearing when one of my kids requests a song

**Reply by gAlleb** [reply]
Thanks @bveenker, was glad to help, maybe we can do something with the `played twice` thingy. I'll ponder.

**Reply by gAlleb** [reply]
Hi @bveenker, I thought how to do it without doubled metadata. Behold another option:

```ruby


# Edit Station API URL (station/{ID_OF_YOUR_STATION}/) and API Key
# Create an API key under My Account → API Keys → Add API Key and copy-paste here.
my_api_url = "http://127.0.0.1/api/station/1/nowplaying"
my_api_key = "99cab3rr2dsvdvds__AN_EXAMPLE__27cbfdc16e178b9c7c"

# Set path to leader's playlist 
#
# (IMPORTANT: Playlist should be advanced with Jingle mode set to true)
# to make it work from the start if your previous playlist wasn't in Jingle mode - you can create fresh new playlist. This way it will be immediately annotated with "jingle_mode".

playlist_request_leader_modified = "/var/azuracast/stations/radio/playlists/your_playlist.m3u"

# Create empty queue
queue_leader = request.queue()

# Now it's track_sensitive=true, because we want it to play before the next song request 
# and it should be higher in a chain
radio = fallback(track_sensitive=true, [queue_leader, radio])

# Function to check API for the next song if its a request or not

def is_next_a_request()
  json_response = ref("")
  def get_api()

   http_response = http.get(my_api_url,
                headers=[
                ("accept", "application/json"),
                ("User-Agent", "Liquidsoap AzuraCast"),
                ("X-API-Key", "#{my_api_key}")
                ],
                timeout=3.0)

    if http_response.status_code != 200 then
      log.important("Request to #{string.quote(my_api_url)} failed with status \
                     #{http_response.status_code}. Status message: \
                     #{string.quote(http_response.status_message)}.")
    null()

    else
      
      try
      	let json.parse ({
        	playing_next
      		} : {
       		playing_next: {
                is_request: bool
                }
                }) = http_response
      "#{playing_next.is_request}"
      catch _ do
          log.important("Unable to parse api response.")
          ""
      end

   end
   
  end


  json_response := get_api()
  #print(json_response())

  if "#{json_response()}" == "true" then

        playlist_lines = playlist.files("#{playlist_request_leader_modified}")
        random_track = list.nth(playlist_lines, random.int(min=0,max=list.length(playlist_lines)))
        queue_leader.push.uri(random_track)

  end

end

# execute function on_track with some delay(5.)
def main(m)

  def null_list(key, _list)
    list.assoc.mem(key, _list) ? list.assoc(key, _list) : "null"
  end

  test = null_list("jingle_mode",m)
  print("#{test}")

  if (m["jingle_mode"] != "true") then

       thread.run(is_next_a_request, delay=5., fast=false)

  end
end

radio.on_metadata(main)
```

If you have `track` -> `jingle` -> `request`. Then your `request_leader` would be played `track` ->  `request_leader` -> `jingle` -> `request`. But I think it's ok.

**Reply by bveenker** [reply]
Wow that's amazing @gAlleb ! I'm a little afraid to try it, the old solution works so perfectly.

I can report that the first solution works with "v0.20.2 Stable" but not on the more recent releases with the AutoDJ scheduler rework. I'm rolling my stations back to v0.20.2 and leaving them there indefinitely. I'll consider upgrading my stations only when this becomes an official feature (and there doesn't seem to be much impetus for that).

**Reply by gAlleb** [reply]
Hi @bveenker, the above new approach works quite fine in the latest version of AC. So don't be hesitant to upgrade :)
